Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp

```diff
@@ -3203,6 +3203,16 @@ void SpeculativeJIT::compileGetByValOnString(Node* node, const ScopedLambda<std:
     std::tie(resultRegs, format, std::ignore) = prefix(node->arrayMode().isOutOfBounds() ? DataFormatJS : DataFormatCell);
     GPRReg scratchReg = resultRegs.payloadGPR();
 
+    if (node->op() == StringAt) {
+        Jump isNotNegativeIndex = branch32(GreaterThanOrEqual, propertyReg, TrustedImm32(0));
+
+        loadPtr(Address(baseReg, JSString::offsetOfValue()), scratchReg);
+        load32(Address(scratchReg, StringImpl::lengthMemoryOffset()), scratchReg);
+        add32(propertyReg, scratchReg, propertyReg);
+
+        isNotNegativeIndex.link(this);
+    }
+
     // unsigned comparison so we can filter out negative indices and indices that are too large
     loadPtr(Address(baseReg, JSString::offsetOfValue()), scratchReg);
     Jump outOfBounds = branch32(
```

Comment:
I think you should add outOfBounds + StringAt + OutOfBounds handling, which just puts undefined to the result.

Reply:
The following code doesn't work (test for outOfBounds fails). Am I doing something wrong?

```cpp
    if (node->op() == StringAt && node->arrayMode().isOutOfBounds()) {
        outOfBounds.link(this);
        moveValue(jsUndefined(), resultRegs);
        doneCases.append(jump());
    }
```